In [1]:
import torch
import numpy as np

In [2]:
labels = torch.tensor([0,1,0,0,0]).float()
preds = torch.tensor([2,1,8,5,6]).float()

In [3]:
BCE = torch.nn.BCEWithLogitsLoss(reduction='none')

## Binary Cross Entropy Loss

In [4]:
BCE(preds, labels)

tensor([2.1269, 0.3133, 8.0003, 5.0067, 6.0025])

In [5]:
label = np.array([0,1,0,0,0])
pred = np.array([2,1,8,5,6])

def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def entropy(x, y):
    return -(x * np.log(sigmoid(y)) + (1-x) * np.log(1-sigmoid(y)))

print(entropy(label, pred))

[2.12692801 0.31326169 8.00033541 5.00671535 6.00247569]


\begin{equation}
loss_1=-[0\times\log{\sigma{(2)}}+(1-0)\times\log({1-\sigma{(2)}})]=-\log({1-\sigma{(2)}})=2.1269 \\
loss_2=-[1\times\log{\sigma{(1)}}+(1-1)\times\log({1-\sigma{(1)}})]=-1\times\log{\sigma{(1)}}=0.3133 \\
loss_3=-[0\times\log{\sigma{(8)}}+(1-0)\times\log({1-\sigma{(8)}})]=-\log({1-\sigma{(8)}})=8.0003 \\
loss_4=-[0\times\log{\sigma{(5)}}+(1-0)\times\log({1-\sigma{(5)}})]=-\log({1-\sigma{(5)}})=5.0067 \\
loss_5=-[0\times\log{\sigma{(6)}}+(1-0)\times\log({1-\sigma{(6)}})]=-\log({1-\sigma{(6)}})=6.0025 \\
\end{equation}

## Cross Entropy Loss

In [6]:
CE = torch.nn.CrossEntropyLoss(reduction='none')
label = torch.tensor([2]).long()
pred = torch.tensor([[2,1,8,5,6]]).float()
CE(pred, label)

tensor([0.1727])

In [7]:
def softmax(x):
    return np.exp(x) / np.exp(x).sum()

pred = np.array([2,1,8,5,6])
label = np.array(2)

pred_softmax = softmax(pred)
print(-np.log(pred_softmax[label]))

0.1727029339230807


$loss=-\log (\frac{e^{8}}{e^2+e^1+e^8+e^5+e^6})=0.1727$

## Negative Log Likelihood Loss

In [10]:
NLL = torch.nn.NLLLoss(reduction='none')
label = torch.tensor([2]).long()
pred = torch.tensor([[2,1,8,5,6]]).float()
NLL(pred, label)

tensor([-8.])

In [13]:
def nllloss(label, pred):
    return -np.log(np.exp(pred[label]))

pred = np.array([2,1,8,5,6])
label = np.array(2)

print(nllloss(label, pred))

-8.0


$loss=-\log (e^8)$

### cross entropy = log softmax + NLLLoss

In [15]:
def log_softmax(x):
    softmax = np.exp(x) / np.exp(x).sum()
    log_softmax = np.log(soft)
    return log_softmax

print(nllloss(label, log_softmax(pred)))

0.1727029339230807
